src/libraries/System.Private.Xml/tests/XmlWriter/XmlTextWriterTests.cs

```diff
@@ -0,0 +1,92 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Collections.Generic;
+using System.IO;
+using Xunit;
+
+namespace System.Xml.Tests
+{
+    public class XmlTextWriterTests
+    {
+        public static IEnumerable<object[]> PositiveTestCases
+        {
+            get
+            {
+                yield return new string[] { null }; // will be normalized to empty string
+                yield return new string[] { "" };
+                yield return new string[] { "This is some data." };
+                yield return new string[] { " << brackets and whitespace >> " }; // brackets & surrounding whitespace are ok
+                yield return new string[] { "&amp;" }; // entities are ok (treated opaquely)
+                yield return new string[] { "Hello\r\nthere." }; // newlines are ok
+                yield return new string[] { "\U0001F643 Upside-down smiley \U0001F643" }; // correctly paired surrogates are ok
+                yield return new string[] { "\uFFFD\uFFFE\uFFFF" }; // replacement char & private use are ok
+            }
+        }
+
+        public static IEnumerable<object[]> BadSurrogateTestCases
+        {
+            get
+            {
+                yield return new string[] { "\uD800 Unpaired high surrogate." };
+                yield return new string[] { "\uDFFF Unpaired low surrogate." };
+                yield return new string[] { "Unpaired high surrogate at end. \uD800" };
+                yield return new string[] { "Unpaired low surrogate at end. \uDFFF" };
+                yield return new string[] { "Unpaired surrogates \uDFFF\uD800 in middle." };
+            }
+        }
+
+        [Theory]
+        [MemberData(nameof(PositiveTestCases))]
+        [InlineData("]]")] // ]] without trailing > is ok
+        [InlineData("-->")] // end of comment marker ok (meaningless to cdata tag)
+        public void WriteCData_SuccessCases(string cdataText)
+        {
```

Comment:
presumably you can use exact same test cases for DataContractSerialization regression?

Reply:
Eh, maybe. The logic is slightly different between the two methods. For example, the DCS methods are intended to be free of side effects if the argument is bad, but only in the comment case - not the cdata case. (I consider this mismatch a bug in the DCS implementation, but whatever.)

Writing a proper test for that would be more complex than what's presented here. We *could* copy this test as-is over to the DCS sources, but I don't think it would provide much value since it's not testing the true intent of the logic. It would also be fragile because it would be testing the logic of *both* `XmlTextWriter` *and* `XmlBaseWriter` simultaneously, which violates the principle that tests should target only a single component.

Given all of that I didn't think it was worthwhile to port the test. But if there's a strong desire to port it, we could.